src/dmd/expressionsem.d

```diff
@@ -7841,6 +7839,19 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
                     if (v && !checkAddressVar(sc, exp.e1, v))
                         return setError();
                 }
+                // https://issues.dlang.org/show_bug.cgi?id=22539
+                if (auto ce = exp.e1.isCallExp())
+                {
+                    if (auto tf = ce.e1.type.isTypeFunction())
+                    {
+                        if (!checkAddressCall(sc, tf))
+                        {
+                            exp.error("cannot slice static array of `ref return` of `%s()` in `@safe` function `%s`",
+                                ce.e1.toChars(), sc.func.toChars());
+                            exp.errorSupplemental("return type `%s` has pointers that may be `scope`", tf.next.toChars());
+                        }
+                    }
+                }
```

Comment:
are identical please put them in a function like you have for `checkAddressCall`

Reply:
I've factored more code out into `checkAddressCall` now